Sync with deps

floragru opened the pull request:
This builds with extructures 0.2.0, coq-utils 0.2.0 (and the math-comp libaries 1.9.0 with coq 8.9.1). Building coq-utils can be streamlined by pinning it with the upstream repo when https://github.com/arthuraa/coq-utils/pull/3 would be merged. In the meantime, `opam pin add coq-utils` and copy-pasting the content of the file added in the coq-utils PR works for me.

floragru committed:
Adapted to coq-utils/extructures v0.2.0 (and mathcomp 1.9)

floragru committed:
Merge branch 'master' into sync_with_deps

floragru committed:
Merge branch 'master' into sync_with_deps

floragru committed:
Merge branch 'master' into sync_with_deps

floragru committed:
Merge branch 'master' into sync_with_deps

floragru committed:
Updated README and dependencies, included coq-utils as submodule

floragru committed:
Added explicit opening of string_scope to patch tests

Incidentally trimmed some lines (automated in my emacs settings)

floragru committed:
dirty patch for tests

(looks like component ids should be coerced to map instead)

floragru commented:
Now the project compiles well with the coq-utils version pinned as a submodule !
Tests are compiling as well but the changes in master that occured in between seem to cause (extremely strangely) a bad derivation of a very simple type (the registers at the intermediate level, which is basically an enum) : 
```
File "./Tests/IntermediateProgramGeneration.v", line 220, characters 0-51:
Error:
Recursive definition of aux_shrink is ill-formed.
...
```
It works well without merging master though

floragru committed:
Merge remote-tracking branch 'origin/master' into sync_with_deps

floragru committed:
Reverted register to an Inductive type for tests

floragru commented:
Ok it's been fixed (register was redefined as a Variant and it didn't work well with QuickChick deriving

floragru commented:
The issue for the derivation of variant is documented here but it doesn't look like it would be taken care of anytime soon https://github.com/QuickChick/QuickChick/issues/155

catalin-hritcu commented:
I'm building Coq directly, never via opam. Can I still build things?

robblanco commented:
@floriangru and I discussed the changes again earlier today. In principle, everything should be ready for a final test and merge; I'll start looking at that now.

@catalin-hritcu As I understand it, you always have the choice of how to build, but you have the option to use `opam` to take care of the dependencies for you. In any case, the OPAM description file does not override the need for a description of the dependencies (if reading them from the file is not good enough).

catalin-hritcu commented on README.md:
```diff
@@ -11,24 +11,52 @@ This repository contains the Coq development of the paper:
-- [CoqUtils](https://github.com/arthuraa/coq-utils/releases/tag/v0.1) 0.1
+This development requires Coq v8.9.1 to work, as well as the following libraries:
+- [Mathematical Components](https://math-comp.github.io/math-comp/) 1.9.0 or dev / on branch master
+- [CoqUtils](https://github.com/arthuraa/coq-utils) on branch master (tested with commit [54c1269](https://github.com/arthuraa/coq-utils/commit/54c1269e1e85e14404e9dab3805e9db448f419f0)),
```
The order of dependencies is wrong here: void and deriving and extstructures are all needed by CoqUtils, so they should appear before it.

catalin-hritcu replied:
Also, what's "dev /" supposed to mean? I've never heard that before.

floragru replied:
> The order of dependencies is wrong here: void and deriving and extstructures are all needed by CoqUtils, so they should appear before it.

Yes indded, since I let opam handle the dependencies I put this one first which fetches all the other ones.

> Also, what's "dev /" supposed to mean? I've never heard that before.

This kind of version is activated if you add the `extra-dev` repo (with some dev versions specific to coq versions, for instance for quickchick, 8.9.dev tracks the 8.9 branch of their repo)

floragru replied:
Changed this bit, silenced some polluting warnings about extraction and with extracted programs when using OCaml >= 4.08.0

catalin-hritcu commented:
This works well for me too ... without any opam beyond for installing OCaml 4.08.1 ... so we should lift the 4.06 upper bound here:

> Our tests are known to work with [QuickChick](https://github.com/QuickChick/QuickChick) branch 8.9 and OCaml from 4.02.3 to 4.06.

Here also the 4.02.3 lower bound seems quite hard to believe ... has anyone tried? :)

floragru committed:
Edited README, bumped most recent OCaml version compatible

oldest one should be 4.05.0 when compatibility with coq 8.10.0 would be
possible

floragru committed:
Silenced warnings with extraction

floragru committed:
Silenced warnings for use of Pervasives (seen in OCaml >= 4.08.0)

Have to see if the extraction behaviour for target program can be
changed to avoid using Pervasives (don't exactly know if it's caused by
QuickChick, coq or our extraction)

cleaned more leftover files in Makefile.Test

floragru committed:
consistency of parameters name in definition and proof, works with coq 8.10.0

In proof mode under coq 8.10.0, these parameters are named according to
the definition (they were strangely renamed in previous versions, with
an appended 0).

Auto trim

floragru commented:
> This works well for me too ... without any opam beyond for installing OCaml 4.08.1 ... so we should lift the 4.06 upper bound here:

Apart from the extracted examples failing for most of them, the testing of SFI worked with 4.09.0 as well (up to 263 generated tests, didn't wait until the 600th)

> > Our tests are known to work with [QuickChick](https://github.com/QuickChick/QuickChick) branch 8.9 and OCaml from 4.02.3 to 4.06.
> 
> Here also the 4.02.3 lower bound seems quite hard to believe ... has anyone tried? :)

We'll have to lift the lower bound to 4.05.0 if we want compatibility with coq 8.10 (which now only depends on Arthur's libraries), so we won't have to try the older versions :relieved:
